src/sqlfluff/core/parser/segments/base.py

```diff
@@ -1266,13 +1296,25 @@ def iter_patches(self, templated_str: str) -> Iterator[FixPatch]:
             # or insert. Also valid if we still have an insertion buffer here.
             end_diff = self.pos_marker.templated_slice.stop - templated_idx
             if end_diff or insert_buff:
-                yield FixPatch(
-                    slice(
-                        self.pos_marker.templated_slice.stop - end_diff,
-                        self.pos_marker.templated_slice.stop,
-                    ),
-                    insert_buff,
+                source_slice = segment.pos_marker.source_slice
+                templated_slice = slice(
+                    self.pos_marker.templated_slice.stop - end_diff,
+                    self.pos_marker.templated_slice.stop,
+                )
+                # By returning an EnrichedFixPatch (rather than FixPatch), which
+                # includes a source_slice field, we ensure that fixes adjacent
+                # to source-only slices (e.g. {% endif %}) are placed
+                # appropriately relative to source-only slices..
+                yield EnrichedFixPatch(
```

Comment:
This feels a little weird, in that we're enriching some patches and not others - but it seems like a clean solution.

I've checked out the flow around it and I think it makes sense. I do wonder whether we should account more explicitly for potential `EnrichedPatch` objects being returned in `fix_string()` where I think in the current flow we might end up recreating them rather than just using the returned ones from `iter_patches`.

btw - I don't think it would be crazy to pass a `TemplatedFile` to `iter_patches` rather than just the `templated_string`. When it's called in `fix_string` it's given the latter from the former anyway. If we did that then you'd have access to the strings so you'd be able to populate the details below.

Reply:
@alanmcruickshank: Done! I also reworked things a bit to make `EnrichedFixPatch` a subclass of `FixPatch`. I think it's significantly cleaner now, and was a useful refactoring independent of the this PR.